src/Microsoft.DotNet.XHarness.CLI/Commands/Android/AndroidRunCommand.cs

```diff
@@ -0,0 +1,261 @@
+
+            // Wait til at least device(s) are ready
+            runner.WaitForDevice();
+
+            // Empty log as we'll be uploading the full logcat for this execution
+            runner.ClearAdbLog();
+
+            logger.LogDebug($"Working with {runner.GetAdbVersion()}");
+
+            return Task.FromResult(InvokeHelper(
+                logger,
+                apkPackageName,
+                _arguments.InstrumentationName,
+                _arguments.InstrumentationArguments,
+                _arguments.OutputDirectory,
+                _arguments.DeviceOutputFolder,
+                _arguments.Timeout,
+                _arguments.ExpectedExitCode,
+                runner));
+        }
+
+        public ExitCode InvokeHelper(
+            ILogger logger,
+            string apkPackageName,
+            string? instrumentationName,
+            Dictionary<string, string> instrumentationArguments,
+            string outputDirectory,
+            string? deviceOutputFolder,
+            TimeSpan timeout,
+            int expectedExitCode,
+            AdbRunner runner)
+        {
+
+            int instrumentationExitCode = (int)ExitCode.GENERAL_FAILURE;
+
+            try
+            {
+                // No class name = default Instrumentation
+                ProcessExecutionResults? result = runner.RunApkInstrumentation(apkPackageName, instrumentationName, instrumentationArguments, timeout);
+                bool processCrashed = false;
+                bool failurePullingFiles = false;
+
+                using (logger.BeginScope("Post-test copy and cleanup"))
+                {
+                    if (result.ExitCode == (int)ExitCode.SUCCESS)
+                    {
+                        (var resultValues, var instrExitCode) = ParseInstrumentationOutputs(logger, result.StandardOutput);
+
+                        // This is where test instrumentation can communicate outwardly that test execution failed
+                        instrumentationExitCode = instrExitCode;
+
+                        // Pull XUnit result XMLs off the device
+                        foreach (string possibleResultKey in s_xmlOutputVariableNames)
+                        {
+                            if (resultValues.ContainsKey(possibleResultKey))
+                            {
+                                logger.LogInformation($"Found XML result file: '{resultValues[possibleResultKey]}'(key: {possibleResultKey})");
+                                try
+                                {
+                                    runner.PullFiles(resultValues[possibleResultKey], outputDirectory);
+                                }
+                                catch (Exception toLog)
+                                {
+                                    logger.LogError(toLog, "Hit error (typically permissions) trying to pull {filePathOnDevice}", resultValues[possibleResultKey]);
+                                    failurePullingFiles = true;
```

Comment:
should we break here? Or do we want to continue with the loop?

Reply:
This is not my code actually, I moved it from `android test`, but I suggest that we should continue because different path may have proper permissions